README.md
# tfann

Tiny Fast Artificial Neural Network Library.

It uses internal tiny SIMD Matrix library.
It can save network structure to file.
It can generate pure dart code with no dependencies from a network.

## Getting Started

typical usage:

```dart
import 'package:tfann/tfann.dart';

...

List<TrainSetInputOutput> xor_data = [
      /*  output: column  1 - XOR of 3 bits, column  2 - AND of 3 bits,
       column  3 - OR of 3 bits, column  4 - if exactly two bits ON,
      */
      TrainSetInputOutput.lists([-1, -1, -1], [0, 0, 0, 0]),
      TrainSetInputOutput.lists([1, 1, -1], [0, 0, 1, 1]),
      TrainSetInputOutput.lists([1, -1, -1], [1, 0, 1, 0]),
      TrainSetInputOutput.lists([-1, 1, -1], [1, 0, 1, 0]),
      TrainSetInputOutput.lists([-1, -1, 1], [1, 0, 1, 0]),
      TrainSetInputOutput.lists([1, 1, 1], [1, 1, 1, 0]),
      TrainSetInputOutput.lists([1, -1, 1], [0, 0, 1, 1]),
      TrainSetInputOutput.lists([-1, 1, 1], [0, 0, 1, 1]),
    ];

final xor_net =
        TfannNetwork.full([3, 4, 4], [ActivationFunctionType.uscsls, ActivationFunctionType.uscsls]);

print("before training...");
xor_data.forEach((data) => print(
        "in: ${data.input.toList()} out: ${xor_net.feedForward(data.input).toList()} expected: ${data.output.toList()}"));

// train network
// train method takes a single TrainSet and runs it only once.
for (int i = 0; i < 10000; ++i) {
      xor_data.forEach((data) {
        xor_net.train(data, learningRate: 0.04);
      });
}


print("after training...");

    
xor_data.forEach((data) => print(
    "in: ${data.input.toList()} out: ${xor_net.feedForward(data.input).toList()} expected: ${data.output.toList()}"));

...


```

To save the network:

```dart
await xor_net.save("binary.net");
```

To load the network:

```dart
var xor_net = TfannNetwork.fromFile("binary.net")!;
```

While developing, use the  --enable-asserts flag, in order to catch bugs.

You may also compile the network into pure dart code. It is very good for production stage.

The produced code have no dependencies at all, even not this package.

Usage:

```dart

print(compileNetwork(xor_net));

```

Output:

```dart
import 'dart:typed_data';
import 'dart:math';

final Float32x4 _SIMD0 = Float32x4.zero();
final Float32x4 _SIMD0_75 = Float32x4.splat(0.75);
final Float32x4 _SIMD0_5 = Float32x4.splat(0.5);
final Float32x4 _SIMD0_25 = Float32x4.splat(0.25);
final Float32x4 _SIMD0_125 = Float32x4.splat(0.125);
final Float32x4 _SIMD0_375 = Float32x4.splat(0.375);
final Float32x4 _SIMD0_625 = Float32x4.splat(0.625);
final Float32x4 _SIMD0_0625 = Float32x4.splat(0.0625);
final Float32x4 _SIMD0_03 = Float32x4.splat(0.03);
final Float32x4 _SIMD0_65625 = Float32x4.splat(0.65625);
final Float32x4 _SIMD0_065 = Float32x4.splat(0.065);
final Float32x4 _SIMD0_185 = Float32x4.splat(0.185);
final Float32x4 _SIMD0_104 = Float32x4.splat(0.104);
final Float32x4 _SIMD0_208 = Float32x4.splat(0.208);
final Float32x4 _SIMD0_704 = Float32x4.splat(0.704);
final Float32x4 _SIMDm0_8 = Float32x4.splat(-0.8);
final Float32x4 _SIMDm1_5 = Float32x4.splat(-1.5);
final Float32x4 _SIMD0_28125 = Float32x4.splat(0.28125);
final Float32x4 _SIMD1 = Float32x4.splat(1);
final Float32x4 _SIMD1_47 = Float32x4.splat(1.47);
final Float32x4 _SIMD1_6 = Float32x4.splat(1.6);
final Float32x4 _SIMD4 = Float32x4.splat(4);
final Float32x4 _SIMD8 = Float32x4.splat(8);
final Float32x4 _SIMDm2 = Float32x4.splat(-2);
final Float32x4 _SIMD0_875 = Float32x4.splat(0.875);
final Float32x4 _SIMD0_4 = Float32x4.splat(0.4);
final Float32x4 _SIMDm0_16 = Float32x4.splat(-0.16);
final Float32x4List _SimdSignMaskVector = Float32x4List.fromList(List.generate(
    16,
    (index) => Float32x4(
        (index & 1) != 0 ? -1.0 : 1.0,
        (index & 2) != 0 ? -1.0 : 1.0,
        (index & 4) != 0 ? -1.0 : 1.0,
        (index & 8) != 0 ? -1.0 : 1.0)));

double uscsls(double x) {
  if (x >= 1.6) return 0.065 * x + 0.704;
  if (x > -0.8) {
    var x2 = x * x;
    var x3 = x2 * x;
    return 0.125 * (x2 - x3) + 0.625 * x;
  }
  return 0.185 * x - 0.208;
}


Float32x4 uscslsX4(Float32x4 x) {
  Int32x4 greater1_6 = x.greaterThan(_SIMD1_6);
  Float32x4 x2 = x * x;

  Float32x4 branch1Result = x.scale(0.065) + _SIMD0_704;
  Float32x4 x3 = x2 * x;

  Int32x4 lessThanMinus0_8 = x.lessThanOrEqual(_SIMDm0_8);
  Float32x4 branch3Result = x.scale(0.185) - _SIMD0_208;

  return greater1_6.select(
      branch1Result,
      lessThanMinus0_8.select(
          branch3Result,
           (x2 - x3).scale(0.125) +  x.scale(0.625)));
}

final List<Float32x4List> Lweight_tfann_evaluate_0 = [Uint32List.fromList([1065924784, 3218828940, 3218824008, 0]).buffer.asFloat32x4List(), Uint32List.fromList([1074832170, 3207276024, 3207270630, 0]).buffer.asFloat32x4List(), Uint32List.fromList([3218045595, 1058827529, 1058838751, 0]).buffer.asFloat32x4List(), Uint32List.fromList([3213025879, 3213257327, 3213261317, 0]).buffer.asFloat32x4List()];
final Float32x4List Lbias_tfann_evaluate_0 = Uint32List.fromList([1051252787, 3212525348, 3213439945, 1049866728]).buffer.asFloat32x4List();
final List<Float32x4List> Lweight_tfann_evaluate_1 = [Uint32List.fromList([3232711821, 1078727539, 3223330061, 1083118854]).buffer.asFloat32x4List(), Uint32List.fromList([3220807383, 3217432562, 3229760405, 3194501247]).buffer.asFloat32x4List(), Uint32List.fromList([3223501112, 1079543989, 1069180988, 3181878151]).buffer.asFloat32x4List(), Uint32List.fromList([1078650051, 1071470358, 1085387923, 3224445642]).buffer.asFloat32x4List()];
final Float32x4List Lbias_tfann_evaluate_1 = Uint32List.fromList([1070831670, 3197145344, 1083611721, 1076128681]).buffer.asFloat32x4List();


List<double> tfann_evaluate(List<double> inData) 
{
  assert(inData.length == 3);
  Float32List input = Float32List(4);
  for (int i = 0; i< 3; ++i) input[i] = inData[i];
  Float32x4List currentTensor = input.buffer.asFloat32x4List();
  Float32List outputTensor;
  outputTensor = Float32List(4);
  for (int r = 0; r < 4; ++r)
  {
    Float32x4List weightRow = Lweight_tfann_evaluate_0[r];
    Float32x4 sum = currentTensor[0]*weightRow[0];
    outputTensor[r] = sum.x + sum.y + sum.z ;
  }
  currentTensor = outputTensor.buffer.asFloat32x4List();
    currentTensor[0]+=Lbias_tfann_evaluate_0[0];
  currentTensor[0]=uscslsX4(currentTensor[0]);
  outputTensor = Float32List(4);
  for (int r = 0; r < 4; ++r)
  {
    Float32x4List weightRow = Lweight_tfann_evaluate_1[r];
    Float32x4 sum = currentTensor[0]*weightRow[0];
    outputTensor[r] = sum.x + sum.y + sum.z + sum.w;
  }
  currentTensor = outputTensor.buffer.asFloat32x4List();
    currentTensor[0]+=Lbias_tfann_evaluate_1[0];
  currentTensor[0]=uscslsX4(currentTensor[0]);
  return currentTensor.buffer.asFloat32List(0,4).toList();
}

```

## Activation functions

Since this library tries the best to utilize your CPU, we propose some new activation functions, that are easy to calculate, and also gives good behavior.

[Abs Sigmoid](https://www.desmos.com/calculator/ukybfmgvot)

[DivLine Sigmoid](https://www.desmos.com/calculator/bvf5vfuola)

[Fast Bell](https://www.desmos.com/calculator/ekmmrldwty)

[Cubic Sigmoid](https://www.desmos.com/calculator/dfnyzvoucc)

[UACSLS](https://www.desmos.com/calculator/ftbmhwspfo)

[USCLS](https://www.desmos.com/calculator/mzjwraayia)

[USCSLS](https://www.desmos.com/calculator/dlfulx2isk)

[Funny Hat](https://www.desmos.com/calculator/idsd8fyrv8)

[All activation functions](https://www.desmos.com/calculator/xwhaw077tn)

## Tips

The train method returns both the forward error (before changing the weights) and the back-propagated error.
You can use the back-propagated error in cases of chaining networks (like RNN or LSTM which not included here).
In these case, you would need to train the network with set of Input and Error. use TrainSetInputError for this case.
You can also use the back-propagated error to create what's called "deep fake" or "deep dream".

If you get NaN in your weights, then you got the infamous exploding gradient problem. Try again and set 'maxErrClipAbove' (one of train arguments- sets a limit for the maximum propagated error) to a small value, or try smaller learning rate.
Also, if you are using unbounded activation functions, try to set few of the layers to bell shaped activation function. The bell functions helps to stabilize the network.


CHANGELOG.md
## 0.6.2+88

* This version breaks compatibiliy with previous versions.
* Modified activation functions, and also new ones.
* New methods in FVector.
## 0.5.2

* "logistic sigmoid" is back to normal, without scaling.
* "compileNetwork" function added instead of TfannNetwork.compile
* New methods in FVector (including array access operator).
* other minor things...
## 0.5.1

* added examples
* enhanced FVector
* changed error calculation
## 0.5.0

* changed the constructor of TfannNetwork

## 0.4.0

* many changes to api
* bug fixes
* more control on training process
* more activation functions

## 0.1.4

* removed flutter dependency

## 0.1.3

* first published package.


lib/src/network.dart
import 'dart:collection';
import 'dart:convert';
import 'dart:typed_data';

import 'package:tfann/tfann.dart';
import 'package:tfann/src/train_set.dart';

import 'activation_function.dart';
import 'dart:math';
import 'dart:io';

import 'linalg.dart';

class RandomSupply {
  static final Random rng = Random();
  static double nextBias() => rng.nextDouble() - 0.5;
  static double nextWeight() {
    var x = rng.nextDouble() * 0.8 + 0.2;
    return rng.nextBool() ? x : -x;
  }
}

class FeedArtifacts {
  final FVector activation;
  final FVector derivative;
  const FeedArtifacts(this.activation, this.derivative);
}

class TrainArtifacts {
  /// defined as Network Output minus Expected Output.
  final FVector forwardError;
  final FVector propagatedError;
  const TrainArtifacts(this.forwardError, this.propagatedError);
}

class TfannLayer {
  FLeftMatrix weights;
  FVector bias;
  ActivationFunction activationFunc;
  TfannLayer.fromJsonMap(Map json)
      : bias = FVector.fromJson(json["bias"]),
        weights = FLeftMatrix.fromJson(json["weights"]),
        activationFunc = mapActivationFunction[
            activationTypeFromString[json["activation"]]!]!;

  TfannLayer(int inputs, int outputs, ActivationFunctionType activationFuncType)
      : bias = FVector.fromList(
            List.generate(outputs, (index) => RandomSupply.nextBias())),
        weights = FLeftMatrix.fromList(List.generate(outputs,
            (o) => List.generate(inputs, (i) => RandomSupply.nextWeight()))),
        activationFunc = mapActivationFunction[activationFuncType]!;

  FVector feedForward(FVector input) {
    return ((weights.multiplyVector(input)) + bias)
      ..apply(activationFunc.func, activationFunc.funcSIMD);
  }

  FeedArtifacts createFeedArtifacts(FVector input) {
    FVector intermediateVector = (weights.multiplyVector(input)) + bias;
    return FeedArtifacts(
        intermediateVector.applied(
            activationFunc.func, activationFunc.funcSIMD),
        intermediateVector.applied(
            activationFunc.derivative, activationFunc.derivativeSIMD));
  }

  Map<String, dynamic> toJson() {
    return {
      "weights": weights.toJson(),
      "bias": bias.toJson(),
      "activation": activationFunc.type.toString()
    };
  }

  int get inputLength => weights.nColumns;
  int get outputLength => weights.nRows;
}

class NetworkFlow {
  NetworkFlow(this.artifacts);
  List<FeedArtifacts> artifacts;
  FVector get output => artifacts.last.activation;
  FVector get input => artifacts.first.activation;
}

/// A structure of a complete fully-connected network.
class TfannNetwork {
  List<TfannLayer> layers = [];
  TfannNetwork(this.layers);
  TfannNetwork.full(
      List<int> layersDefinition, List<ActivationFunctionType> activation) {
    assert(layersDefinition.length == activation.length + 1);

    for (int i = 0; i < activation.length; ++i) {
      layers.add(TfannLayer(
          layersDefinition[i], layersDefinition[i + 1], activation[i]));
    }
  }

  /// Get the network's output vector from the input vector.
  FVector feedForward(FVector input) {
    assert(input.length == layers.first.inputLength);
    return layers.fold(input, (vec, layer) => layer.feedForward(vec));
  }

  /// Propagates forward the input vector, and collects the outputs and derivatives of each layer.
  ///
  /// The first element in the resulted list contains the input vector along with a zeros vectors that reflects the derivatives.
  /// Each element afterwards contains the output vector and derivative vector of the n'th layer.
  NetworkFlow createNetworkFlow(FVector input) {
    List<FeedArtifacts> artifacts = [
      FeedArtifacts(input, FVector.zero(input.length))
    ];
    FVector currentInput = input;
    for (TfannLayer l in layers) {
      FeedArtifacts currentArtifact = l.createFeedArtifacts(currentInput);
      artifacts.add(currentArtifact);
      currentInput = currentArtifact.activation;
    }
    return NetworkFlow(artifacts);
  }

  /// Train network with a single training pair, for a single epoch.
  ///
  /// Returns the propagated error of the first layer, which is good for chained networks.
  TrainArtifacts train(TrainSet trainSet,
      {double learningRate = 0.04,
      double maxErrClipAbove = 0.0,
      double skipIfErrBelow = 0.0,
      bool Function(FVector)?
          skipIfOutput, //TODO: replace with error weight vector or something?
      NetworkFlow? networkFlow}) {
    assert(trainSet.input.length == layers.first.inputLength);
    //create the input and output of each layer, and the derivatives
    if (networkFlow == null) {
      networkFlow = createNetworkFlow(trainSet.input);
    }

    FVector netOutput = networkFlow.output;
    List<FeedArtifacts> artifacts = networkFlow.artifacts;

    FVector netErrors;

    /// Normalize/clip the error if needed...
    if (trainSet is TrainSetInputOutput) {
      assert(trainSet.output.length == layers.last.outputLength);
      netErrors = netOutput - trainSet.output;
    } else {
      assert((trainSet as TrainSetInputError).error.length ==
          layers.last.outputLength);
      netErrors = (trainSet as TrainSetInputError).error;
    }
    if (skipIfOutput?.call(netOutput) ?? false) {
      return TrainArtifacts(netErrors, FVector.zero(layers.first.inputLength));
    }
    FVector normalizedErrors = netErrors;
    if (maxErrClipAbove > 0.0) {
      double norm = normalizedErrors.abs().largestElement();
      if (norm < skipIfErrBelow) {
        return TrainArtifacts(
            netErrors, FVector.zero(layers.first.inputLength));
      }
      //double norm = normalizedErrors.squared().sumElements();
      if (norm > maxErrClipAbove) {
        normalizedErrors = netErrors.scaled(maxErrClipAbove / norm);
      }
    }

    //Do the actual training...

    FVector previousDelta = normalizedErrors;
    List<FVector> layerDelta = [];
    for (int i = layers.length - 1; i >= 0; --i) {
      FVector currentDelta = (artifacts[i + 1].derivative * previousDelta);
      layerDelta.add(currentDelta);

      previousDelta =
          (layers[i].weights.transposed().multiplyVector(currentDelta));
    }

    var arti = artifacts.iterator;
    for (TfannLayer l in layers) {
      l.bias -= layerDelta.last.scaled(learningRate);
      arti.moveNext();

      l.weights -= layerDelta.last.multiplyTransposed(arti.current.activation)
        ..scale(learningRate);
      layerDelta.removeLast();
    }
    return TrainArtifacts(netErrors, previousDelta);
  }

  /// Given a single training pair, calculate the network's mean-square-error.
  FVector calculateMeanSquareError(List<TrainSetInputOutput> data) {
    var sumVector = FVector.zero(layers.last.outputLength);
    data.forEach((trainSet) {
      sumVector += (feedForward(trainSet.input) - trainSet.output).squared();
    });
    return sumVector..scale(1.0 / data.length);
  }

  /// Given a single training pair, calculate the network's mean-absolute-error per output node
  FVector calculateMeanAbsoluteError(List<TrainSetInputOutput> data) {
    var sumVector = FVector.zero(layers.last.outputLength);
    data.forEach((trainSet) {
      sumVector += (feedForward(trainSet.input) - trainSet.output).abs();
    });
    return sumVector..scale(1.0 / data.length);
  }

  /// Returns Json object.
  ///
  /// To convert it to Json string, use 'jsonEncode'
  Object toJson() {
    return layers.map((l) => l.toJson()).toList();
  }

  /// Saves the network to file.
  Future<void> save(String filename) async {
    var jsonString = jsonEncode(toJson());
    var sink = File(filename).openWrite(mode: FileMode.writeOnly);
    sink.write(jsonString);

    // Close the IOSink to free system resources.
    await sink.flush();
    sink.close();
  }

  static TfannNetwork? fromFile(String filename) {
    String jsonString = File(filename).readAsStringSync();
    return TfannNetwork.fromJson(jsonDecode(jsonString));
  }

  /// Creates TfannNetwork from Json object.
  ///
  /// Use 'jsonDecode' to convert a Json into Object.
  factory TfannNetwork.fromJson(dynamic jsonObject) {
    assert(jsonObject is List);
    assert((jsonObject as List).isNotEmpty);
    if (jsonObject is List) {
      assert(jsonObject.every((element) => element is Map));
      return TfannNetwork(
          jsonObject.map((e) => TfannLayer.fromJsonMap(e)).toList());
    }
    throw ArgumentError(
        "cannot construct a TfannNetwork from this json object");
  }
}
